### PR TITLE
spec:ABCI2.0 - Uncommenting content related to vote xtensions and finalize block

### LINKED ISSUE #403: Uncomment all parts related to vote extensions, and `FinalizeBlock` from the spec.
Uncomment and adjust the spec in main to address `FinalizeBlock` and `VoteExtensions`

### PR DESCRIPTION
**Pull request by jmalicevic**
The PR will be a draft until PR #407  is merged. After that it will be rebased to `feature/abci++vef` .

Closes #403 .

### CODE REVIEW

File: spec/abci/abci++_app_requirements.md
Comment (lasarojc) [inline spec/abci/abci++_app_requirements.md:93]:
```suggestion
in round *r*, height *h*.
```

File: spec/abci/abci++_app_requirements.md
Comment (lasarojc) [inline spec/abci/abci++_app_requirements.md:441]:
```suggestion
`FinalizeBlock` is the workhorse of the blockchain. CometBFT delivers the decided block,
```

**Review by lasarojc — APPROVED** [review]
Some minor comments.
Some suggestions that may be incorporated in this PR or maybe directly on main, in PR #433

**Review by lasarojc — CHANGES_REQUESTED** [review]
Some minor comments.
Some suggestions that may be incorporated in this PR or maybe directly on main, in PR #433

File: spec/abci/abci++_methods.md
Comment (cason) [inline spec/abci/abci++_methods.md:365]: This <tab> here makes GitHub to format the remaining of this item list as a code. The comments, in particular, are rendered :D

File: spec/abci/abci++_methods.md
Comment (cason) [inline spec/abci/abci++_methods.md:695]: `CommitInfo` is used in ProcessProposal and FinalizeBlock.

**Review by cason — APPROVED** [review]
I have left some comments, they should not block merging the PR, in particular since the most relevant were addressed by recent commits.

File: spec/abci/abci++_basic_concepts.md
Comment (jmalicevic) [reply spec/abci/abci++_basic_concepts.md:68]: It's now below in finalize block.

File: spec/abci/abci++_comet_expected_behavior.md
Comment (jmalicevic) [reply spec/abci/abci++_comet_expected_behavior.md:61]: I think it's because this is a non-proposer? @sergio-mena

File: spec/abci/abci++_methods.md
Comment (jmalicevic) [reply spec/abci/abci++_methods.md:333]: Hmm this is how it was in the 0.36 spec. I will check commits onto 0.37 to see whether the `next_validator_hash` info was added in that branch. I think it should be included in the sentence.

Comment (sergio-mena) [reply spec/abci/abci++_methods.md:333]: OK, after some digging, I found out what's going on.
In #98, I was cherry-picking code that might have been forgotten in `v0.36.x`, that we might need. It was that PR that introduced `next_validators_hash`.
I didn't backport #98 to `v0.37.x` because the initial PR had to do with ABCI 2.0. Although I realize I should backport it. Will open a PR shortly.
Thanks for spotting @cason !

File: spec/abci/abci++_methods.md
Comment (sergio-mena) [inline spec/abci/abci++_methods.md:369]: I understand your hesitation.
We modified this late in the `v0.37.x` process following SDK's problems with the previous interface, so we won't change it now.
Conclusion, we need to keep the text that says that there are no additional checks.

File: spec/abci/abci++_methods.md
Comment (sergio-mena) [reply spec/abci/abci++_methods.md:695]: It's part of
* `CommitInfo` in `ProcessProposal` and `FinalizeBlock`
* `ExtendedCommitInfo` in `PrepareProposal`

File: spec/abci/abci++_app_requirements.md
Comment (sergio-mena) [inline spec/abci/abci++_app_requirements.md:37]:
```suggestion
as a form of immediate execution.
```

File: spec/abci/abci++_app_requirements.md
Comment (sergio-mena) [inline spec/abci/abci++_app_requirements.md:211]:
```suggestion
decided block's data to the Application, which uses it to transition its state, but MUST NOT persist it;
persisting MUST be done during `Commit`.
```
I'm suggesting this now because this helps in tackling #203

File: spec/abci/abci++_app_requirements.md
Comment (sergio-mena) [inline spec/abci/abci++_app_requirements.md:365]:
```suggestion
results MUST respect the same order as the list of transactions delivered via
```

**Review by sergio-mena — APPROVED** [review]
Thanks Jasmina! I know it mustn't have been obvious to reconcile both versions, given that they are sprinkled with small changes we made on-the-go over months.

File: spec/abci/abci++_methods.md
Comment (sergio-mena) [reply spec/abci/abci++_methods.md:333]: #453

File: spec/abci/abci++_app_requirements.md
Comment (lasarojc) [reply spec/abci/abci++_app_requirements.md:37]: I believe that optimistic is more accurate here, because it signals that the execution may be wasted, while immediate does not. It is also a term commonly used for similar execution of transactions in DBs. I understand that we have used immediate in some other places so far, though, so would need to review them.

File: spec/abci/abci++_app_requirements.md
Comment (lasarojc) [reply spec/abci/abci++_app_requirements.md:34]: I'd suggest `safety` instead of correctness, then, as using the wrong timeout cannot cause processes to disagree or violate validity. `Liveness`, however, seems incorrect as timeouts can lead to infinitely many extra rounds in which nil is pre-voted.

Comment (lasarojc) [inline spec/abci/abci++_app_requirements.md:34]: Discussing with @sergio-mena , we reached the following alternative text for line 34 (GitHub wouldn't let me add it as a suggestions.

```suggestion
Note that the violation of Requirement 1 may lead to further rounds, but will not compromise liveness because even though `TimeoutPropose` is used as the initial value for proposal timeouts, CometBFT will be dynamically adjust these timeouts such that they will eventually be enough for completing `PrepareProposal`.
```

Comment (lasarojc) [reply spec/abci/abci++_app_requirements.md:34]: Discussing with @sergio-mena , we reached the following alternative text for **line 34** (GitHub wouldn't let me add it as a suggestion).

```
Note that the violation of Requirement 1 may lead to further rounds, but will not 
compromise liveness because even though `TimeoutPropose` is used as the initial
value for proposal timeouts, CometBFT will be dynamically adjust these timeouts
such that they will eventually be enough for completing `PrepareProposal`.
```

File: spec/abci/abci++_basic_concepts.md
Comment (jmalicevic) [reply spec/abci/abci++_basic_concepts.md:44]: I left this as `created` but removed the `,`

File: spec/abci/abci++_methods.md
Comment (jmalicevic) [reply spec/abci/abci++_methods.md:333]: I will merge this w/o this fix then ,since you opened a PR and then we can merge the PR into this.

File: spec/abci/abci++_app_requirements.md
Comment (jmalicevic) [reply spec/abci/abci++_app_requirements.md:37]: I tend to agree with Lasaro, immediate sounds like the execution is final and the state change is persisted.

**Comment by jmalicevic** [comment]
Thank you @sergio-mena , @cason and @lasarojc  for reviewing. All the comments that were not applied (most of all Daniel's suggestions for text reformatting and the discussion on optimistic vs. immediate execution, I merged in a doc so we can get back to this later) but I will merge this as to unblock other minor PRs regarding the spec.

**Pull request merged**
